### PR TITLE
Add methods and classes for Places (New) API in `flutter_google_places_sdk_platform_interface`

### PR DESCRIPTION
**Pull request by flikkr**
### Description

Add classes and methods introduced in Google Places (New) API. I have some changes to add to Android library once this is merged. Tracked by #77.

### CODE REVIEW

File: flutter_google_places_sdk_platform_interface/analysis_options.yaml [hunk 1]
```diff
@@ -48,7 +48,6 @@ linter:
     - always_declare_return_types
     - always_put_control_body_on_new_line
     # - always_put_required_named_parameters_first # we prefer having parameters in the same order as fields https://github.com/flutter/flutter/issues/10219
-    - always_require_non_null_named_parameters
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/analysis_options.yaml:51]: don't change the analysis options

File: flutter_google_places_sdk_platform_interface/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,11 @@
+## 0.3.0
+
+* BREAKING: Rename `RankPreference` to `TextSearchRankPreference` and introduce `NearbySearchRankPreference`
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/CHANGELOG.md:3]: RankReference doesn't exists so the "Rename" change doesn't make sense

File: flutter_google_places_sdk_platform_interface/lib/flutter_google_places_sdk_platform.dart [hunk 3]
```diff
@@ -9,18 +9,19 @@ abstract class FlutterGooglePlacesSdkPlatform extends PlatformInterface {
 
   /// "Powered by google" image that should be used when background is white
   static const AssetImage ASSET_POWERED_BY_GOOGLE_ON_WHITE = AssetImage(
-      'assets/google/powered_by_google_on_white.png',
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/flutter_google_places_sdk_platform.dart:12]: don't change reformatting either

File: flutter_google_places_sdk_platform_interface/lib/src/types/lat_lng_bounds.dart [hunk 4]
```diff
@@ -22,3 +22,17 @@ class LatLngBounds with _$LatLngBounds {
   factory LatLngBounds.fromJson(Map<String, Object?> json) =>
       _$LatLngBoundsFromJson(json);
 }
+
+/// An immutable class representing a latitude/longitude aligned circle, with a defined radius.
+@Freezed()
+class CircularBounds with _$CircularBounds {
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/lat_lng_bounds.dart:28]: keep this in it's own file

File: flutter_google_places_sdk_platform_interface/lib/src/types/author_attribution.dart [hunk 5]
```diff
@@ -0,0 +1,24 @@
+
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'author_attribution.freezed.dart';
+part 'author_attribution.g.dart';
+
+@freezed
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/author_attribution.dart:7]: use `@Freezed()`

File: flutter_google_places_sdk_platform_interface/lib/src/types/place.dart [hunk 6]
```diff
@@ -39,6 +41,7 @@ class Place with _$Place {
     required int? utcOffsetMinutes,
     required LatLngBounds? viewport,
     required Uri? websiteUri,
+    required List<Review>? reviews,
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/place.dart:44]: add these to the changelog

File: flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart [hunk 7]
```diff
@@ -143,8 +143,103 @@ enum PlaceType {
   TRAVEL_AGENCY,
   UNIVERSITY,
   VETERINARY_CARE,
-  ZOO
-  ;
+  ZOO,
+  /// Places (new) types
+  AMERICAN_RESTAURANT,
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart:148]: perhaps we should have a new enum? or these types can be used with the old api as well?

File: flutter_google_places_sdk_platform_interface/lib/src/types/place_type_filter.dart [hunk 8]
```diff
@@ -58,5 +57,6 @@ enum PlaceTypeFilter {
 }
 
 extension PlaceTypeFilterDescriptor on PlaceTypeFilter {
+  @Deprecated('Use `name` instead')
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/place_type_filter.dart:60]: If we should use name, we can just return name in the value and note it in the changelog; no need to deprecate it.

File: flutter_google_places_sdk_platform_interface/lib/src/types/review.dart [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  /// Constructs a [Review] object.
+  const factory Review({
+    /// The attribution that must be shown to the user if this review is displayed.
+    required String attribution,
+
+    /// This review's AuthorAttribution.
+    required AuthorAttribution authorAttribution,
+
+    /// The text of the review in its original language.
+    String? originalText,
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/review.dart:18]: put required fields first, than others after

File: flutter_google_places_sdk_platform_interface/lib/src/types/review.dart [hunk 10]
```diff
@@ -0,0 +1,41 @@
+import 'package:flutter_google_places_sdk_platform_interface/src/types/author_attribution.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'review.freezed.dart';
+part 'review.g.dart';
+
+@freezed
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/review.dart:7]: `@Freezed()`

File: flutter_google_places_sdk_platform_interface/lib/src/types/search_by_text_response.dart [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter_google_places_sdk_platform_interface/flutter_google_places_sdk_platform_interface.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'search_by_text_response.freezed.dart';
+
+/// The response for a [FlutterGooglePlacesSdkPlatform.searchByText] request
+@freezed
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/search_by_text_response.dart:7]: `@Freezed()`

File: flutter_google_places_sdk_platform_interface/lib/src/types/search_by_text_response.dart [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+class SearchByTextResponse with _$SearchByTextResponse {
+  /// constructs a [SearchByTextResponse] object.
+  const factory SearchByTextResponse(
+    /// the Place list returned by the response.
+    List<Place> places,
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/search_by_text_response.dart:12]: required

File: flutter_google_places_sdk_platform_interface/lib/src/types/search_nearby_response.dart [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter_google_places_sdk_platform_interface/flutter_google_places_sdk_platform_interface.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'search_nearby_response.freezed.dart';
+
+/// The response for a [FlutterGooglePlacesSdkPlatform.searchNearby] request
+@freezed
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/search_nearby_response.dart:7]: `@Freezed()`

File: flutter_google_places_sdk_platform_interface/lib/src/types/search_nearby_response.dart [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+class SearchNearbyResponse with _$SearchNearbyResponse {
+  /// constructs a [SearchNearbyResponse] object.
+  const factory SearchNearbyResponse(
+    /// the Place list returned by the response.
+    List<Place> places,
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/lib/src/types/search_nearby_response.dart:12]: required

File: flutter_google_places_sdk_platform_interface/pubspec.yaml [hunk 15]
```diff
@@ -11,7 +11,7 @@ dependencies:
     sdk: flutter
   freezed_annotation: ^2.4.1
   json_annotation: ^4.8.1
-  plugin_platform_interface: ^2.1.6
+  plugin_platform_interface: ^2.1.8
```
Comment (matanshukry) [inline flutter_google_places_sdk_platform_interface/pubspec.yaml:14]: I'm assuming it's not needed for this change, so let's separate it into another pr; this one is big as it is

**Comment by matanshukry** [comment]
hey @flikkr , thanks for the PR!
There was a discussion not too long ago: https://github.com/matanshukry/flutter_google_places_sdk/issues/65
The reason I'm not so sure about moving to the new api is that least comment.

If you can comment on that with some sources from the documentation that would be great - otherwise I would want to do more research before deciding to move.

Regardless:
1. I've added a couple of review, take a look when you can
2. If you can also create a PR on how you use it - that would be great. That 2nd pr doesn't have to be perfect nor even build, but it's hard to comment on this without knowing how it will even be used.

**Comment by flikkr** [comment]
Hi @matanshukry, thanks for taking the time to look into this! I've created a sample PR #91 to show how I integrate the full feature in my project.

Regarding the concern on moving to the new API, there should not be any breaking changes as the new API can only be used once it is enabled via the Google Cloud console and initialised in app by calling [`Places.initializeWithNewPlacesApiEnabled`](https://developers.google.com/maps/documentation/places/android-sdk/config#connect-client). The old API can still be used by calling `Places.initialize`. ([see migration guide](https://developers.google.com/maps/documentation/places/android-sdk/migrate-overview))

File: flutter_google_places_sdk_platform_interface/lib/src/types/search_by_text_response.dart (hunk 12 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/lib/src/types/search_by_text_response.dart:12]: This is not a named parameter so does not need `require` keyword

File: flutter_google_places_sdk_platform_interface/lib/src/types/place.dart (hunk 6 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place.dart:44]: It's already there

File: flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart (hunk 7 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart:148]: They cannot be used in the old API, however using them in the old API will not throw any exception and the place type will just be ignored. IMO, keeping these new types in the current enum is fine, consumers will just need to be aware that these types are only compatible with the new API as mentioned in the Places SDK docs.

File: flutter_google_places_sdk_platform_interface/analysis_options.yaml (hunk 1 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/analysis_options.yaml]: Done a395bbb8e8df6257f8cfb0c8f5dc25291b1d98c3

File: flutter_google_places_sdk_platform_interface/lib/src/types/place_type_filter.dart (hunk 8 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place_type_filter.dart]: Done a395bbb8e8df6257f8cfb0c8f5dc25291b1d98c3

File: flutter_google_places_sdk_platform_interface/lib/src/types/search_nearby_response.dart (hunk 14 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/lib/src/types/search_nearby_response.dart:12]: This is not a named parameter so does not need require keyword

File: flutter_google_places_sdk_platform_interface/pubspec.yaml (hunk 15 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/pubspec.yaml]: Done a395bbb8e8df6257f8cfb0c8f5dc25291b1d98c3

File: flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart (hunk 7 above)
Comment (matanshukry) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart:148]: > consumers will just need to be aware that these types are only compatible with the new AP

@flikkr not a fan of that; we want to make it as easy as possible for users (consumers), so they'll make as little errors as possible

**Comment by matanshukry** [comment]
@flikkr again sorry for the long delays;
Anyway let's just create a separate enum for the new api 🙏 , and we can merge this after

File: flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart (hunk 7 above)
Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart:148]: @matanshukry I don't think this is an unreasonable approach. This is basically what Google is doing in their own SDKs as well. They basically combine both new and existing place fields in the same enum, leaving it to the developer to ensure proper usage.

Comment (matanshukry) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart:148]: @flikkr just because someone else is doing it doesn't make it right - I'm into logic, not blindly following others.

I don't understand why we would want to do that; what's the advantage of having 1 enum, now giving another responsibility to our users, rather than limit it on our own and making it clear for them

Comment (flikkr) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart:148]: @matanshukry How do you see this being implemented? If we decide on a separate enum, we would need to add this as a parameter to every function that currently accepts a `PlaceField`. Taking the `fetchPlace` function as an example, we need to ensure the user is passing at least 1 field to the function across these 2 parameters, one of which is tagged as `required`. 

```dart
// Holds all the 'new' place fields
enum PlaceFieldNew {
  ...
}

// Existing API
@override
Future<FetchPlaceResponse> fetchPlace(
  String placeId, {
  required List<PlaceField> fields, // <= Pass existing place fields here
  List<PlaceFieldNew>? newFields, // <= Pass all 'new' place fields here
  bool? newSessionToken,
}) {
   ...
}
```

It is also possible that the user may want to pass a 'new' field and not an existing field (e.g. I only want **place reviews**), in this case the user would need to pass an empty list to the `fields` param

```dart
fetchPlace(
  fields: [],
  newFields: [PlaceFieldNew.Reviews]
);
```

Comment (matanshukry) [reply flutter_google_places_sdk_platform_interface/lib/src/types/place_type.dart:148]: @flikkr 

tl&dr: merging this PR.

-------

Solution #1: in the actual api implementations we'll probably need both, but we can wrap it in a single object with 2 fields.
For the user, we can use Freezed with union types. Something like
```
@Freezed()
class PlaceField with _$PlaceField {
  const factory PlaceField.v1(PlaceFieldV1 v1) = _PlaceFieldV1;
  const factory PlaceField.v2(PlaceFieldV2 v2) = _PlaceFieldV2;
}
```

and later use getters to send both

-------

After the above solution, I figured if there are more changes that are only applicable to v2, the even better solution would be to have an entire strong type class dedicated for v2. Something like:

```
class FlutterGooglePlacesSdk {
  FlutterGooglePlacesSdkNew initializeNew(..) { .. }
  FlutterGooglePlacesSdkOld initialize(..) { .. }
}

class FlutterGooglePlacesSdkCommon {
  .. updateSettings(..) { .. }
}

class FlutterGooglePlacesSdkNew extends FlutterGooglePlacesSdkCommon {
  type newMethod(..) { implement }
}

class FlutterGooglePlacesSdkOld extends FlutterGooglePlacesSdkCommon {
  type oldMethod(..) { implement
}
```

that way you can initialize old/new, and there's no way you'll accidentally use an old/new method that you're not allowed to.

-------

And finally, I realized:
1. Google marks this old as deprecated, and hopefully it won't be relevant at all soon enough - so not worth working on it too much
2. I'm not highly maintaining this package, and I can't expect others to do that amount of work

**Comment by matanshukry** [comment]
@flikkr I fixed the tests and workflow in mater, after resolving conflicts it's now missing the `return null` statement.
Let's fix it and make sure the tests passes, and merge this

**Comment by flikkr** [comment]
Hi @matanshukry I fixed the failing test, please help to run the workflow
